File: backend/app/api/analytics.py
"""Analytics API endpoints."""
from typing import Optional
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.api.dependencies import get_current_employer, get_current_admin
from app.services.analytics import get_analytics_service
from app.services.subscription import get_tier_limits
from app.models.employer import Employer, SubscriptionTier

router = APIRouter(tags=["analytics"])


@router.get("/analytics/scraping")
async def get_scraping_metrics(
    source_platform: Optional[str] = None,
    days: int = 7,
    db: Session = Depends(get_db),
    admin: dict = Depends(get_current_admin),
):
    """Get scraping metrics (admin only)."""
    analytics_service = get_analytics_service(db)
    metrics = analytics_service.get_scraping_metrics(
        source_platform=source_platform,
        days=days,
    )
    return metrics


@router.get("/analytics/searches/popular")
async def get_popular_searches(
    limit: int = 10,
    db: Session = Depends(get_db),
):
    """Get popular search terms."""
    analytics_service = get_analytics_service(db)
    searches = analytics_service.get_popular_searches(limit=limit)
    return {"popular_searches": searches}


@router.get("/analytics/jobs/{job_id}")
async def get_job_analytics(
    job_id: UUID,
    days: int = 30,
    db: Session = Depends(get_db),
    employer: Employer = Depends(get_current_employer),
):
    """Get analytics for a specific job (premium tier only)."""
    tier_limits = get_tier_limits(employer.subscription_tier)
    
    if not tier_limits.get("has_analytics", False):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Analytics access requires premium subscription tier",
        )
    
    from app.models.job import Job
    job = db.query(Job).filter(Job.id == job_id).first()
    if not job:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Job not found",
        )
    
    if job.employer_id != employer.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You do not have permission to view analytics for this job",
        )
    
    analytics_service = get_analytics_service(db)
    analytics = analytics_service.get_job_analytics(job_id=job_id, days=days)
    return analytics


@router.get("/analytics/employer/summary")
async def get_employer_analytics_summary(
    days: int = 30,
    db: Session = Depends(get_db),
    employer: Employer = Depends(get_current_employer),
):
    """Get aggregate analytics for all employer's jobs (premium tier only)."""
    tier_limits = get_tier_limits(employer.subscription_tier)
    
    if not tier_limits.get("has_analytics", False):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Analytics access requires premium subscription tier",
        )
    
    analytics_service = get_analytics_service(db)
    analytics = analytics_service.get_employer_analytics(
        employer_id=employer.id,
        days=days,
    )
    return analytics


@router.get("/analytics/system/health")
async def get_system_health_metrics(
    metric_name: Optional[str] = None,
    hours: int = 24,
    db: Session = Depends(get_db),
    admin: dict = Depends(get_current_admin),
):
    """Get system health metrics (admin only)."""
    analytics_service = get_analytics_service(db)
    metrics = analytics_service.get_system_metrics(
        metric_name=metric_name,
        hours=hours,
    )
    return {"metrics": metrics}


@router.get("/analytics/api/slow-requests")
async def get_slow_api_requests(
    threshold_ms: float = 1000,
    hours: int = 24,
    limit: int = 100,
    db: Session = Depends(get_db),
    admin: dict = Depends(get_current_admin),
):
    """Get slow API requests (admin only)."""
    analytics_service = get_analytics_service(db)
    requests = analytics_service.get_slow_requests(
        threshold_ms=threshold_ms,
        hours=hours,
        limit=limit,
    )
    return {"slow_requests": requests}

File: frontend/lib/api/analytics.ts
import apiClient from '@/lib/api-client'

export interface ScrapingMetric {
  source_platform: string
  total_tasks: number
  successful_tasks: number
  failed_tasks: number
  jobs_found: number
  jobs_created: number
  jobs_updated: number
  success_rate: number
  avg_duration_seconds: number
}

export interface PopularSearch {
  query: string
  count: number
}

export interface JobAnalytics {
  job_id: string
  view_count: number
  application_count: number
  views_by_day: Array<{ date: string; count: number }>
  applications_by_day: Array<{ date: string; count: number }>
}

export interface EmployerAnalyticsSummary {
  total_jobs: number
  active_jobs: number
  expired_jobs: number
  filled_jobs: number
  total_views: number
  total_applications: number
  avg_applications_per_job: number
  top_performing_jobs: Array<{
    job_id: string
    title: string
    view_count: number
    application_count: number
  }>
}

export interface SystemHealthMetric {
  metric_name: string
  value: number
  timestamp: string
  status: 'healthy' | 'warning' | 'critical'
}

export interface SlowAPIRequest {
  endpoint: string
  method: string
  duration_ms: number
  timestamp: string
  status_code: number
}

export const analyticsApi = {
  // Get scraping metrics
  getScrapingMetrics: async (sourcePlatform?: string): Promise<ScrapingMetric[]> => {
    const response = await apiClient.get<ScrapingMetric[]>('/analytics/scraping', {
      params: { source_platform: sourcePlatform },
    })
    return response.data
  },

  // Get popular searches
  getPopularSearches: async (limit: number = 10): Promise<PopularSearch[]> => {
    const response = await apiClient.get<PopularSearch[]>('/analytics/searches/popular', {
      params: { limit },
    })
    return response.data
  },

  // Get job analytics
  getJobAnalytics: async (jobId: string): Promise<JobAnalytics> => {
    const response = await apiClient.get<JobAnalytics>(`/analytics/jobs/${jobId}`)
    return response.data
  },

  // Get employer analytics summary
  getEmployerAnalyticsSummary: async (days: number = 30): Promise<EmployerAnalyticsSummary> => {
    const response = await apiClient.get<EmployerAnalyticsSummary>('/analytics/employer/summary', {
      params: { days },
    })
    return response.data
  },

  // Get system health metrics (admin only)
  getSystemHealthMetrics: async (metricName?: string): Promise<SystemHealthMetric[]> => {
    const response = await apiClient.get<SystemHealthMetric[]>('/analytics/system/health', {
      params: { metric_name: metricName },
    })
    return response.data
  },

  // Get slow API requests (admin only)
  getSlowAPIRequests: async (thresholdMs: number = 1000, limit: number = 50): Promise<SlowAPIRequest[]> => {
    const response = await apiClient.get<SlowAPIRequest[]>('/analytics/api/slow-requests', {
      params: { threshold_ms: thresholdMs, limit },
    })
    return response.data
  },
}
